src/net/p2p.rs

```diff
@@ -3028,11 +3050,355 @@ impl PeerNetwork {
+                        // we'd have skipped this step.
+                        warn!("Unusable Data URL {}", url_str);
+                        self.mempool_sync_reset();
+                        return Ok(true);
+                    }
+                }
+            }
+        } else {
+            // can't do anything
+            debug!("No DNS client, and URL contains a domain, so no mempool sync can happen");
+            self.mempool_sync_reset();
+            return Ok(true);
+        }
+    }
+
+    /// Ask the remote peer for its mempool, connecting to it in the process if need be.
+    /// sets self.mempool_sync_event_id
+    /// Returns Ok(true) if we're done mempool syncing
+    /// Returns Ok(false) if there's more to do
+    fn mempool_sync_send_query(
+        &mut self,
+        url: UrlString,
+        addr: SocketAddr,
+        mempool: &MemPoolDB,
+        chainstate: &mut StacksChainState,
+    ) -> Result<bool, net_error> {
+        let sync_data = mempool.make_mempool_sync_data()?;
+        let request = HttpRequestType::MemPoolQuery(
+            HttpRequestMetadata::from_host(PeerHost::from_socketaddr(&addr)),
+            sync_data,
+        );
+
+        let event_id =
+            self.connect_or_send_http_request(url, addr, request, mempool, chainstate)?;
+        self.mempool_sync_event_id = Some(event_id);
+        return Ok(false);
+    }
+
+    /// Receive the mempool sync response.
+    /// Return Ok(true) if we're done with the mempool sync.
+    /// Return Ok(false) if we have more work to do.
+    fn mempool_sync_recv_response(&mut self, event_id: usize) -> Result<bool, net_error> {
+        PeerNetwork::with_http(self, |network, http| {
+            match http.get_conversation(event_id) {
+                None => {
+                    if http.is_connecting(event_id) {
+                        debug!("Mempool sync event {} is not connected yet", event_id,);
+                        return Ok(false);
+                    } else {
+                        // conversation died
+                        debug!("Mempool sync peer hung up");
+                        network.mempool_sync_reset();
+                        return Ok(true);
+                    }
+                }
+                Some(ref mut convo) => {
+                    match convo.try_get_response() {
+                        None => {
+                            // still waiting
+                            debug!(
+                                "Mempool sync event {} still waiting for a response",
+                                event_id
+                            );
+                            return Ok(false);
+                        }
+                        Some(http_response) => match http_response {
+                            HttpResponseType::MemPoolTxs(_, txs) => {
+                                network.mempool_sync_transactions = Some(txs);
```

Comment:
Is it possible to just return the synced transactions rather than adding them to the network state? It seems like this caller's caller is checking the `Option<>` field after every invocation to shuttle the data out into the network result object. It should be possible (and would simplify the data flow in the code) to just return this data instead, right?

Reply:
Sure; refactored